### ISSUE TITLE
Add additional fold variants to ZStream

### ISSUE THREAD

**Issue by iravid** [zio/zio#1463]
Add additional `fold` variants to `ZStream`:
```scala
// These exist:
// A fold that can exit early
final def fold[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZIO[R1, E1, S]
// A fold that can exit early, that carries the stream's scope
final def foldManaged[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZManaged[R1, E1, S]
// A pure fold that consumes everything
final def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]

// These don't:
// An effectful fold that consumes everything
final def foldLeftM[R1 <: R, E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => ZIO[R1, E1, S]): ZIO[R1, E1, S]
// A pure fold that consumes everything and carries the stream's scope
final def foldLeftManaged[A1 >: A, S](s: S)(f: (S, A1) => S): ZManaged[R, E, S]
// An effectful fold that consumes everything and carries the stream's scope
final def foldLeftManagedM[R1 <: R, E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => ZIO[R1, E1, S]): ZManaged[R1, E1, S]
```

**Comment by vasilmkd** [comment]
Let's first decide on the naming scheme.

**Comment by kamilkloch** [comment]
How should it differ from the readily available `ZStream#foldManaged`?

**Comment by iravid** [comment]
The signatures are:
```scala
// These exist:
// A fold that can exit early
final def fold[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZIO[R1, E1, S]
// A fold that can exit early, that carries the stream's scope
final def foldManaged[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZManaged[R1, E1, S]

// These don't:
// A pure fold that consumes everything
final def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]
// An effectful fold that consumes everything
final def foldLeftM[R1 <: R, E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => ZIO[R1, E1, S]): ZIO[R1, E1, S]
// A pure fold that consumes everything and carries the stream's scope
final def foldLeftManaged[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]
// An effectful fold that consumes everything and carries the stream's scope
final def foldLeftManagedM[R1 <: R, E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => ZIO[R1, E1, S]): ZManaged[R1, E1, S]
```

**Comment by iravid** [comment]
@kamilkloch see above :-)

**Comment by kamilkloch** [comment]
Thanks the extra info. 
How these two should differ in implementation? The signatures are the same. 

> ```scala
> // These exist:
> // A pure fold that consumes everything
> final def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]
> 
> // These don't:
> // A pure fold that consumes everything and carries the stream's scope
> final def foldLeftManaged[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]
> ```

**Comment by iravid** [comment]
Sorry, the second one should return managed.

**Comment by kamilkloch** [comment]
@jdegoes @regiskuckaertz @vasilmkd what is your take on the naming scheme here?
Not entirely sure about `fold(Left)(Managed)(M)`. 

Current status:
```scala
// A fold that can exit early
final def fold[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZIO[R1, E1, S]
// A fold that can exit early, that carries the stream's scope
final def foldManaged[R1 <: R, E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => ZIO[R1, E1, S]): ZManaged[R1, E1, S]
// A pure fold that consumes everything
final def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S]
```

Proposition:
- drop `Left` suffix,
- rename `fold`s with termination condition to `foldWhile`,
- append `M` to effectful versions

Possible variants:

```scala
def fold(s: S)(f: (S, A1) => S): ZIO  // standard signature
def foldM(s: S)(f: (S, A1) => ZIO): ZIO
def foldWhile(s: S)(cont: S => Boolean)(f: (S, A1) => S): ZIO
def foldWhileM(s: S)(cont: S => Boolean)(f: (S, A1) => ZIO): ZIO
def foldManaged(s: S)(f: (S, A1) => S): ZManaged
def foldManagedM(s: S)(f: (S, A1) => ZIO): ZManaged
def foldWhileManaged(s: S)(cont: S => Boolean)(f: (S, A1) => S): ZManaged
def foldWhileManagedM(s: S)(cont: S => Boolean)(f: (S, A1) => ZIO): ZManaged
```

**Comment by vasilmkd** [comment]
@kamilkloch I like your proposal.

**Comment by kamilkloch** [comment]
Please take a look. An awful lot of combinators.

**Comment by ghostdogpr** [comment]
Closed by #1756
